### ISSUE TITLE
RHash - build MSYS and CygWin issues including patch for 1.3.5

### ISSUE THREAD

**Issue by JPeterMugaas** [rhash/RHash#48]
[001-mingw-cygwin-msys.patch.txt](https://github.com/rhash/RHash/files/1364241/001-mingw-cygwin-msys.patch.txt)

I was compiling RHash with MSYS and GNU GCC 6.30 as a prerequisit for cmake and I ran into some issues tried to address.  The patch does work with RHash 1.3.5 (but I'm not sure about the master branch).  There were several issues that I tried to fix.

1) MSYS2 and Cygwin do NOT use the lib prefix name for the .DLL.  The conventions are the prefix "msys-" for MSYS and "cyg" for CygWin.  I had to make a way to override that in the makefile with the variable "DLLNAME_PREFIX".
2) I wanted the make file to do some processing specifically for Windows such as have a separate .DLL and .dll.a import library for shared builds as well as build .EXE's using that convention.  I'm well aware that GCC can use the imports right from the .DLL but this is not often done (probably for compatibility with compilers that are not able to support this.  Thus, I also created another variable for IS_WIN that supports this as well as triggers some things with tests in Windows so that you could have a more uniform usage in the Makefile.  The most ideal  fix would be to consider doing something like use autoconf or cmake.
3) Cygwin and MSYS2 use the standard " __declspec(dllexport)" just like Windows.  Those are just software layers that sit above Windows.  Along this line, you can actually use some Windows functions within MSYS for a half-in-half out approach.
4) When using the SSL runtime, the OpenSSL names are different in MSYS and CYGWIN.  In MSYS, it's probably "msys-crypto-1.0.0.dll" and in CygWin, it's "cygcrypto-1.0.0.dll".
5) MSYS GCC defines "__MSYS__" in addition to "__CYGWIN__".  CygWin defines "__CYGWIN__".  _WIN32 is NOT defined for MSYS and CygWin.  This is something that has to be noted when reviewing my work as I deliberately omitted "__MSYS__".
6) I have had to use the frequency timer stuff from Windows since some tests depended upon "set_benchmark_cpu_affinity" and I also thought it might be better for the static build tests and perhaps for RHash.exe if it was using that capability.
7) In "RHash-1.3.5/tests/test_rhash.sh", I noticed that the "//" path issue ONLY applies if you are compiling and testing the native Windows version of the MINGW RHash.  When building and testing the "MSYS" version for the MSYS environment, you don't get that issue so I had to adjust the workaround in the .sh script appropriately.  Note that I was using the MSYS2 (http://www.msys2.org/) pacman "makepkg" facility including the "makepkg-mingw" for MINGW builds.

I have verified that the patch does work in MSYS2 as well for MINGW-64 as well as MINGW32.

**Comment by JPeterMugaas** [comment]
I have gone ahead and did a patch for the version in GIT that seems to work.  I decided to rework another of things in the make file as well as added some other nice goodies such as:

1) pkg-config support
2) There are now three different sets of build and install rules.  I felt that this was clearner than the old IS_WIN variable and the NAME prefix variable.
3) I experimented with the Python bindings and it turns out to work if you use the proper run-time library names.
4) /etc is no longer hard-coded in the Makefile and the parse_cmd.c file.  I added a macro to the versions file that the Makefile creates. You can now override /etc with something like SYSCONFDIR=/mingw64/etc .  This should help with some atypical file placements in some Unix systems and also assists me with a relocation that I do in a different patch for Windows.
5) A separate importlib is now generated for MINGW, MSYS, and Cygwin builds.  I realize that GCC could directly use the DLL for imports but this is not typical and the filename for the DLL Will be different in those environments.

Please feel free to use this patch and provide me with any feedback. 
[001-mingw-cygwin-msys.patch.txt](https://github.com/rhash/RHash/files/1376355/001-mingw-cygwin-msys.patch.txt)

**Comment by rhash** [comment]
Hello.

First, thank you for writing good patch!

I will start incorporate your changes. But I will not use GNU make specific code (ifneq) because it is not working from scratch on BSD/Solaris/other Unix. For difficult configuration detection I'm going to write configure script (without using autoconf). It's expected to proceed slowly, because of limited time I have for the project, but I consider the configure script very important, particularly for future features ;)

**Comment by JPeterMugaas** [comment]
I think I make right a cmake patch if you wish or possibly an autoconf.

**Comment by JPeterMugaas** [comment]
Alright, I have been spending quite a bit of time on a build system that uses CMake and I think that might be a good option considering that you don't like Autoconf and because CMake will be more portable.

CMake works in three ways with the build system.
1) It generates the appropriate Makefiles for the particular system and adjustments for many environments.
2) You can have the build system such as GNU make call CMake to run cmake scripts that do things such as run programs, perform tests, and some other stuff.
3) You could use cmake command from the command line to do things such as delete files, make symlinks, copy files, or even use something like echo.

Part of this is also that we could reduce dependencies on various Unix bashisms and small utilities such as grep, tr, and sed since Windows does typically include those.  This is taking a bit of time because there is a lot to do but I think it would be work doing considering that CMake itself uses the RHash library.

**Comment by rhash** [comment]
@JPeterMugaas the most of your patch was incorporated into the master source tree by commit cd560d9c3ddc8c91912ea5e965261e27ea056da8

Could you please test it on MSYS (or Cygwin)?

P.S. There is unsolved issue about install-implib and install-so-link targets - each distro has different rules what libs/links to put into different packages.

**Comment by JPeterMugaas** [comment]
Quite frankly, I’m seeing a lot of problems that I have to look at.  

 

There still is a lot to do and I have not gotten around to an issue with test pathes and separators.  The issue there is that unlike with regular Win32 applications, the path separator is the same in Cygwin and MSYSS as in Unix so your “hack” workaround does NOT work.

 

Now some things I need to explain are these:

1.	Cygwin and MSYS sit on top of Windows and are actually similar in many respects.  They all rely on .DLL’s that are put in the /bin path.  In addition, exported functions in .DLL’s are some done the same way as in standard Windows.
2.	MSYS and CYGWIN both use __CYGWIN__  as a define.  For something MSYS specific, you use __MSYS__ .  Note that _WIN32 is not defined for both MSYS2 and CygWin except in older compilers.
3.	The prefix for the .DLL itself is “msys-“ in MSYS2.  In CygWin, it’s “cyg”.  But the imported library and static objects use “lib”as the prefix just like in Mingw.
4.	In MSYS and CygWin and I suspect, MINGW, libiconv is used for the iconv functions.  In Linux, it’s done with the libc library or something else.
5.	In MSYS and Cygwin as well as ming, the OpenSSL crypto library uses “crypto” for the lib name for linking.  It may be the DLL name itself that uses the easy*.dll name or the “crypto*” name.
6.	In Windows, the .DLL’s (shared objects) are placed in the bin directory, not lib.  The linking is typically done through the imported library, .dll.a which is in the lib directory.

 

As I say,  some of this is still not finished because of a linking problem I’m having.

 

Right now, I’m dumbfounded by this:

 

/usr/lib/../lib/libmsys-2.0.a(libcmain.o): In function `main':

/msys_scripts/msys2-runtime/src/msys2-runtime/winsup/cygwin/lib/libcmain.c:37: undefined reference to `WinMain'

/msys_scripts/msys2-runtime/src/msys2-runtime/winsup/cygwin/lib/libcmain.c:37:(.text.startup+0x94): relocation truncated to fit: R_X86_64_PC32 against undefined symbol `WinMain'

**Comment by rhash** [comment]
I see, the scripts still should be polished.

The WinMain error probably appears because -shared flags is no passed in $SHARED_LDFLAGS to the dll linking command:
$(CC) $(SHARED_CFLAGS) $(SOURCES) $(SHARED_LDFLAGS) -o $@

Try compiling the latest master branch, a felt I fixed something like this last days.

**Comment by JPeterMugaas** [comment]
Alright, I'm doing some more refinements on the build system right now but I have made some progress.

**Comment by JPeterMugaas** [comment]
Alright, I have taken the liberty of working further on the build and I’ve got many of the issues solved with it and I’m attaching a patch.

 

There are some unusual things about Win32 builds including Cygwin and msys as I’ve said so I had to make ifeq’s in the make file and put a few more vars in the configuration system.

 

All .DLL’s must go into the bin dir.  The reason is so that the .EXE can load them.  In addition, I also had to fix a problem with symlinks.  The .EXE’s in windows have the .exe extension that needed to be tacked on.

 

This also will install the import library for the .DLL.

 

PS, I’m also including a patch for the ruby bidings to ensure that a file closed before it’s deleted.

 

It might be interesting to see if I can actually get the configuration script to work with the bindings as well so that the bindings could be packaged.  Mingw does have support for ruby, python, and possibly mono.  MSYS has support for perl and python.
